internal/pkg/provision/providers/firecracker/create.go

```diff
@@ -28,80 +27,77 @@ func (p *provisioner) Create(ctx context.Context, request provision.ClusterReque
+	if err == nil {
+		return nil, fmt.Errorf(
+			"state directory %q already exists, is the cluster %q already running? remove cluster state with osctl cluster destroy",
+			state.StateDirectory,
+			request.Name,
+		)
 	}
 
-	httpServer, err := inmemhttp.NewServer(fmt.Sprintf("%s:0", request.Network.GatewayAddr))
-	if err != nil {
-		return nil, err
+	if !os.IsNotExist(err) {
+		return nil, fmt.Errorf("error checking state directory: %w", err)
 	}
 
-	for _, node := range request.Nodes {
-		var cfg string
-
-		cfg, err = node.Config.String()
-		if err != nil {
-			return nil, err
-		}
-
-		if err = httpServer.AddFile(fmt.Sprintf("%s.yaml", node.Name), []byte(cfg)); err != nil {
-			return nil, err
-		}
+	err = os.MkdirAll(state.StateDirectory, os.ModePerm)
+	if err != nil {
+		return nil, fmt.Errorf("error creating state directory: %w", err)
 	}
 
-	state.baseConfigURL = fmt.Sprintf("http://%s/", httpServer.GetAddr())
-
-	httpServer.Serve()
-	defer httpServer.Shutdown(ctx) //nolint: errcheck
+	fmt.Fprintln(options.LogWriter, "creating network", request.Network.Name)
 
-	state.tempDir, err = ioutil.TempDir("", "talos")
-	if err != nil {
-		return nil, err
+	if err = p.createNetwork(ctx, state, request.Network); err != nil {
+		return nil, fmt.Errorf("unable to provision CNI network: %w", err)
 	}
 
-	fmt.Fprintf(options.LogWriter, "created temporary environment in %q\n", state.tempDir)
-
 	var nodeInfo []provision.NodeInfo
 
 	fmt.Fprintln(options.LogWriter, "creating master nodes")
 
-	if nodeInfo, err = p.createNodes(ctx, state, request, request.Nodes.MasterNodes()); err != nil {
+	if nodeInfo, err = p.createNodes(state, request, request.Nodes.MasterNodes()); err != nil {
 		return nil, err
 	}
 
 	fmt.Fprintln(options.LogWriter, "creating worker nodes")
 
 	var workerNodeInfo []provision.NodeInfo
 
-	if workerNodeInfo, err = p.createNodes(ctx, state, request, request.Nodes.WorkerNodes()); err != nil {
+	if workerNodeInfo, err = p.createNodes(state, request, request.Nodes.WorkerNodes()); err != nil {
 		return nil, err
 	}
 
 	nodeInfo = append(nodeInfo, workerNodeInfo...)
 
-	// TODO: temporary, need to wait for all nodes to finish bootstrapping
-	//       before shutting down config HTTP service
-	time.Sleep(30 * time.Second)
-
-	state.clusterInfo = provision.ClusterInfo{
+	state.ClusterInfo = provision.ClusterInfo{
 		ClusterName: request.Name,
 		Network: provision.NetworkInfo{
-			Name: request.Network.Name,
-			CIDR: request.Network.CIDR,
+			Name:        request.Network.Name,
+			CIDR:        request.Network.CIDR,
+			GatewayAddr: request.Network.GatewayAddr,
+			MTU:         request.Network.MTU,
 		},
 		Nodes: nodeInfo,
 	}
 
-	return state, nil
+	// save state
+	stateFile, err := os.Create(filepath.Join(state.StateDirectory, stateFileName))
+	if err != nil {
+		return nil, err
+	}
+
+	defer stateFile.Close() //nolint: errcheck
+
+	if err = yaml.NewEncoder(stateFile).Encode(&state); err != nil {
```

Comment:
What is the benefit of `Encode` vs `Unmarshal`?

Reply:
functionally equivalent, but `Encode` works on `io.Writer` directly without buffering in memory. no big deal for a small structure here, but a big difference for huge objects.